session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift

```diff
@@ -0,0 +1,231 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+
+internal typealias DiffableID = Int64
+
+/// A base interface of array elements compared in `computeDiff(oldArray:newArray:)`.
+internal protocol Diffable {
+    /// Unique identifier of elements in `oldArray` and `newArray`.
+    ///
+    /// It is used to determinen the type of mutation applied to the element:
+    /// - "add" - if certain `id` exists in `newArray`, but not in `oldArray`,
+    /// - "remove" - if certain `id` exists in `oldArray`, but not in `newArray`,
+    /// - "update" - if certain `id` exists in both arrays and `isDifferent(than:)` returns `true`.
+    var id: DiffableID { get }
+
+    /// Determines if this element has any changes if compared to other element.
+    /// It is used from `computeDiff(oldArray:newArray:)`. Both elements are guaranteed to have the same `id`.
+    /// - Parameter otherElement: the element to compare against
+    /// - Returns: `true` if data in both elements is different; `false` otherwise
+    func isDifferent(than otherElement: Self) -> Bool
+}
+
+/// Describes differences between both arrays compared in `computeDiff(oldArray:newArray:)`. It lists minimal
+/// set of changes that needs to be applied to `oldArray` to result with `newArray`.
+///
+/// Note:
+/// - when reconstructing `newArray` from `oldArray` and `Diff` **removals must be applied before additions**;
+/// - if certain element changes its position in `newArray`, `Diff` includes its **removal and addition** (on new position).
+internal struct Diff<Element: Diffable> {
+    /// Dictates that new element was added in `newArray`.
+    struct Add {
+        /// The `id` of preceding element in `newArray` or `nil` if new element should be inserted at index `0`.
+        let previousID: DiffableID?
+        /// The element to add.
+        let new: Element
+    }
```

Comment:
As discussed, it's gonna be `nextID: DiffableID?` in place of `previousID`. It's not a big change (and no change to the algorithm itself). It can be updated in another PR.